File: app/src/main/java/com/roy/jump/util/AndroidUtil.java
package com.roy.jump.util;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.widget.Toast;

import com.roy.jump.App;

/**
 * Created by Roy on 2022/2/19
 */
public class AndroidUtil {
  private static final DisplayMetrics dm = Resources.getSystem().getDisplayMetrics();
  private static final int screenWidth = dm.widthPixels;
  private static final int screenHeight = dm.heightPixels;
  private static Toast toast = null;

  public static int dp(final float dp) { return (int) (0.5f + dp * dm.density); }

  public static int sp(final float dp) { return (int) (0.5f + dp * dm.scaledDensity); }

  public static int getScreenWidth() { return screenWidth; }

  public static int getScreenHeight() { return screenHeight; }

  public static int getStatusBarHeight() {
    final Resources resources = getApplicationContext().getResources();
    final int resourceId = resources.getIdentifier("status_bar_height", "dimen", "android");
    return resourceId > 0 ? resources.getDimensionPixelSize(resourceId) : 0;
  }

  public static Context getApplicationContext() {
    return App.getApp().getApplicationContext();
  }

  public static String getPackageName() {
    return getApplicationContext().getPackageName();
  }

  public static int getColor(final int resId) {
    final Context context = getApplicationContext();
    return context.getResources().getColor(resId);
  }

  public static String getString(final int resId) {
    final Context context = getApplicationContext();
    return context.getResources().getString(resId);
  }

  public static void toast(final String text) {
    if (!TextUtils.isEmpty(text)) {
      if (null != toast) {
        toast.cancel();
      }
      if (null == toast) {
        toast = Toast.makeText(getApplicationContext(), text, Toast.LENGTH_SHORT);
      } else {
        toast.setText(text);
      }
      toast.show();
    }
  }

  public static Drawable makeShape(final int solidColor,
                                   final float cornerRadius,
                                   final int strokeColor,
                                   final float strokeWidth) {
    final GradientDrawable drawable = new GradientDrawable();
    drawable.setColor(solidColor);
    drawable.setCornerRadius(dp(cornerRadius));
    drawable.setStroke(dp(strokeWidth), strokeColor);
    return drawable;
  }

  public static void runOnUIThread(final Runnable runnable) {
    runOnUIThread(runnable, 0);
  }

  public static void runOnUIThread(final Runnable runnable, final long delay) {
    final Handler handler = App.getHandler();
    if (null == handler || null == runnable) {
      return;
    }
    if (0 >= delay) {
      if (Looper.getMainLooper() == Looper.myLooper()) {
        runnable.run();
      } else {
        handler.post(runnable);
      }
    } else {
      handler.postDelayed(runnable, delay);
    }
  }

}

File: app/src/main/java/com/roy/jump/util/AndroidUtil.kt
package com.roy.jump.util

import android.content.ComponentName
import android.content.Context
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.text.TextUtils
import com.roy.jump.App
import com.roy.jump.R
import com.roy.jump.service.PerformService

/**
 * Created by Roy on 2021/9/11
 */
object AndroidUtil {
  private val handler = Handler(Looper.getMainLooper())
  fun runOnUIThread(runnable: Runnable) {
    runOnUIThread(0, runnable)
  }

  fun runOnUIThread(delay: Long, runnable: Runnable) {
    if (0 >= delay) {
      if (Looper.myLooper() == Looper.getMainLooper()) {
        runnable.run()
      } else {
        handler.post(runnable)
      }
    } else {
      handler.postDelayed(runnable, delay)
    }
  }

  /**
   * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
   *
   * @param dp float
   * @return int dpValue对应的px值
   */
  fun dp(dp: Float): Float {
    val scale = App.appCtx.resources.displayMetrics.density
    return (dp * scale + 0.5f)
  }

  fun createShape(radius: Float, solidColor: Int, strokeColor: Int, strokeWidth: Int): Drawable {
    val shape = GradientDrawable()
    shape.cornerRadius = radius
    shape.setStroke(strokeWidth, strokeColor)
    shape.setColor(solidColor)
    return shape
  }

  fun getColor(colorId: Int): Int {
    return App.appCtx.resources.getColor(colorId, null)
  }

  fun isAccessibilityServiceEnabled(): Boolean {
    val context = App.appCtx
    val expectedComponentName = ComponentName(context, PerformService::class.java)
    val enabledServicesSetting = Settings.Secure.getString(context.contentResolver, Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES) ?: return false
    val colonSplitter = TextUtils.SimpleStringSplitter(':')
    colonSplitter.setString(enabledServicesSetting)
    while (colonSplitter.hasNext()) {
      val componentNameString = colonSplitter.next()
      val enabledService = ComponentName.unflattenFromString(componentNameString)
      if (enabledService != null && enabledService == expectedComponentName) {
        return true
      }
    }
    return false
  }
}